File: two-pointers/0015-3sum/solution.java
// 15. 3Sum
// https://leetcode.com/problems/3sum/
// Medium | Java | Accepted 2024-09-23
// Runtime 836 ms | Memory 52.2 MB

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int target = 0;
        Arrays.sort(nums);
        Set<List<Integer>> s = new HashSet<>();
        List<List<Integer>> output = new ArrayList<>();
        for (int i = 0; i < nums.length; i++){
            int j = i + 1;
            int k = nums.length - 1;
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum == target) {
                    s.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    j++;
                    k--;
                } else if (sum < target) {
                    j++;
                } else {
                    k--;
                }
            }
        }
        output.addAll(s);
        return output;
    }
}

File: two-pointers/0015-3sum/solution.cpp
// 15. 3Sum
// https://leetcode.com/problems/3sum/
// Medium | C++ | Accepted 2026-07-31
// Runtime 47 ms | Memory 29 MB

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> sol;
        sort(nums.begin(), nums.end());
        int k = nums.size()-1;
        for(int i = 0; i<nums.size()-2; i++)
        {
            if(i>0 && nums[i-1]==nums[i])
            {
                continue;
            }
            int j = i+1;
            k = nums.size()-1;
            while(j<k)
            {
                if(nums[i] + nums[j] + nums[k]>0)
                {
                    k--;
                }
                else if(nums[i]+nums[j]+nums[k]<0)
                {
                    j++;
                }
                else 
                {
                    sol.push_back(vector<int>{nums[i], nums[j], nums[k]});
                    while(j<k && j+1<nums.size() && nums[j+1]==nums[j])
                    {
                        j++;
                    }
                    j++;
                    while(j<k && k>i && nums[k-1] == nums[k])
                    {
                        k--;
                    }
                    k--;
                }
            }
        }
        return sol;
    }
};
